README.md
# Python integration with the Verbio Speech Center cloud.

This repository contains a python example of how to use the Verbio Technologies Speech Center cloud both for speech recognition and speech synthesis, allowing users to test the streaming capabilities of Speech Center. 

## Index

1. Pre-Requirements
2. Set-up
3. Speech recognition
4. Speech Synthesis
5. Customer Credentials
6. Dashboard


## Pre-Requirements
Before starting you will need:

1. Dashboard user account (provided to you by Verbio Technologies), please contact our sales team at info@verbio.com to get one.
2. Speech center streaming endpoint valid for speech recognition and speech synthesis. Please check the endpoints section.
3. Speech Center proto files (provided under proto directory in this repository).
4. Python 3.6+
5. Protobuf package 3.20.x or lower (or set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python but this will use pure-Python parsing and will be much slower).
6. Pip

## Endpoints

Speech Center streaming endpoint for STT (Speech-To-Text) and TTS (Text-To-Speech):  **us.speechcenter.verbio.com**

##  Set-up
The steps needed are very similar to the ones described in the grpc official guide.

### Install dependencies
You can use the standard pip call to install all the necessary dependencies, from the root of the repo run:
```shell
pip install -r requirements.txt
```

### Generate grpc code with python
In scripts directory there is a `generate_grpc_code.sh` script that will generate the gRPC and Protobuf code for you.
```shell
cd scripts/
./generate_grpc_code.sh
```

The script must be called from inside the `scripts` directory. It will generate all needed grpc files on the project root directory `proto/generated` like:

```text
verbio_speech_center_recognizer_pb2.py
verbio_speech_center_recognizer_pb2_grpc.py
...
```

### Run a client

The CLI clients will use the generated code to connect to the speech center cloud to process your speech file. To run the clients, run either the `recognizer_stream.py` for CSR requests, or the `synthesizer.py` for TTS requests, from the `cli-client` directory. Check the speech recognition or speech synthesis sections of this readme for more information.


## Speech-To-Text

Speech Center allows to easily convert an audio resource into its associated text. Access the `cli-client` directory to run `recognizer_stream.py` located inside to test Speech Center's speech recognition (CSR) capabilities. Before proceeding, the client needs either a valid token for Speech Center or the user's client credentials. Please check the **Customer Credentials** section of this README before proceeding if you haven't got either.

Speech recognition CLI will retrieve an audio file stored locally and send it through streaming to Speech Center's GRPC interface. To run the client please run the `recognizer_stream.py` from the `cli-client` directory. An example transcription request is shown below, all the flags are then detailed.


**Example for streaming**

```shell
cd cli-client/
python3 recognizer_stream.py --audio-file file.wav --topic GENERIC --language en-US --host us.speechcenter.verbio.com --token token.file --asr-version V1 --label project1
```

You can use the `--help`command for more options.

This code will generate the following terminal output on success:
```text
[2023-04-04 12:28:29,078][INFO]:Running speechcenter streaming channel...
[2023-04-04 12:28:29,080][INFO]:Connecting to us.speechcenter.verbio.com
[2023-04-04 12:28:29,082][INFO]:Running executor...
[2023-04-04 12:28:29,083][INFO]:Sending streaming message config
[2023-04-04 12:28:29,083][INFO]:Running response watcher
[2023-04-04 12:28:29,083][INFO]:Waiting for server to respond...
[2023-04-04 12:28:29,084][INFO]:Sending streaming message audio
[2023-04-04 12:28:29,084][INFO]:All audio messages sent
[2023-04-04 12:31:27,109][INFO]:New incoming response: '{  "result": {    "alternatives": [      {     ...'
	"transcript": "Hi. My name is John Doe.",
	"confidence": "0.899752",
	"duration": "4.460000"
[2023-04-04 12:31:27,110][INFO]:New incoming response: '{  "result": {    "alternatives": [      {     ...'
	"transcript": "I'd like to check my account balance, please.",
	"confidence": "0.994662",
	"duration": "7.000000"
[2023-04-04 12:31:32,111][INFO]:Stream inactivity detected, closing stream...
[2023-04-04 12:31:32,112][INFO]:Recognition finished


```

You can also run:
```shell
python3 recognizer_stream.py --help
```  
to list all the available options.

### Client flags for Speech-To-Text (STT)


#### Audio file

Argument: 
```
-a, --audio file
```

This argument is required, stating a path to a .wav audio in 8kHz or 16kHz sampling rate and PCM16 encoding to use for the recognition.

#### Topic

```
-T, --topic arg
```

 Topic to use for the recognition when a grammar is not provided. Must be `GENERIC` | `BANKING` | `TELCO` | `INSURANCE` (default: `GENERIC`).
 
> **THIS FEATURE IS STILL IN DEVELOPMENT, PLEASE USE THE GENERIC TOPIC FOR ALL REQUESTS OR AN ERROR WILL BE GIVEN.**

#### Language

```
-l, --language arg
```

Language to use for the recognition: `en-US`, `en-GB`, `pt-BR`, `es`, `es-419`, `tr`, `ja`, `fr`, `fr-CA`, `de`, `it` (default: `en-US`). Please check the [Speech Center streaming Soeech-To-Text documentation](https://speechcenter.verbio.com/documentation/index.php/speech-to-text-streaming/) for an updated list of language codes.

#### Diarization

```
-d, --diarization
```

This option enables diarization.

> This option is oriented towards batch transcription only and its use for streaming and call automation is not recommended.


#### Formatting

```
-f, --formatting
```

This option will enable formatting on the speech transcription. Please check the [Speech Center streaming Speech-To-Text documentation](https://speechcenter.verbio.com/documentation/index.php/speech-to-text-streaming/) for an updated list of languages that accept formatting.

#### Speech-To-Text engine version

```
-A, --asr-version arg
```

This will select the Speech-To-Text engine version the speech center will use for transcriptions.

> Please follow Verbio's sales department recommendation on which version to use.

#### Sample rate


```
-s, --sample-rate arg
```

Sampling rate for the audio generation: 8000 or 16000.

#### Token


```
-t, --token arg 
```

Path to the authentication token file. This file needs to have a valid token in order for the Speech Center to work. 

In order for the client to work, the token argument is required in the following situations:

- The client will authenticate just by using the available token file. The file provided in this argument **needs to be a valid SpeechCenter JWT token so the transcription service can work**.
- The client will authenticate by providing their client credentials through the `--client_id` and `--client_secret` program arguments. In this case **a token file must also be supplied even if it is a blank file**. Client will check file to see if it is a valid token, if it isn't it will refresh automatically the token and fill the file with a valid token. In this case, client_id and client_secret fields are also required.


#### Client id and secret


```
--client-id arg      Client id for token refresh (default: "")
```    

``` 
--client-secret arg  Client secret for token refresh (default: "")
``` 

`client-id` and `client-secret` fields are required for automatic token refreshal. The arguments need to be written inline with no quotes for each field. Please see the '**Automatically Refresh Service Token**' section above for specific details.

#### Host


```
-H, --host arg
```

URL of the host where the request will be sent. Main endpoints will be expanded as the product is deployed in different regions. Please use **us.speechcenter.verbio.com** as the host for CSR requests.

#### Labels

```
-L, --label arg
```

This option allows for a one word argument to be sent so that the speech transcription request is billed as part of a particular project for the customer. The argument will be a one word name that will classify the request under that project.

- **Project name must only consist of one word.**
- **Argument must be the same each time for the same project. If there is a typo another project will be created.**
- **There is no limit on the amount of projects that can be created.**


## Text-To-Speech

Verbio's Speech Center allows to synthesize text in a whole catalogue of voices. Access the `cli-client` directory to run `synthesizer.py` located inside to test Speech Center's Text-To-Speech (TTS) capabilities. Before proceeding, the client needs either a valid token for Speech Center or the user's client credentials. Please check the **Customer Credentials** section of this README before proceeding if you haven't got either.

**Example for text to speech**
Synthesize request using valid token:
```
python3 synthesizer.py --text "Hello, my friend!" --voice tommy_en_us --audio-file out.wav --host your_speechcenter_hostname --token ./my.token
```
Using client credentials for automated token refreshal:
```
python synthesizer.py --text "Hello my friend!" --voice tommy_en_us --audio-file out.waw --host your_speechcenter_ hostname --client-id your_client_id --client-secret your_client_secret --token token.file
```

Please check out documentation at speechcenter.verbio.com/documentation for current TTS capabilities.

### Text-To-Speech Client flags

#### Help

Argument: 
```
-h, --help
```
Help menu detailing the client's options and a list of all available voices.

#### Text

```
-T, --text "My text string"
```

String of text to convert into speech.


#### Voice

```
-v, --voice arg
```

Voice to use for the text to speech generation. Please bear in mind that each voice has also the language code encoded in it. Some of the available voices are `tommy_en_us` (US English),`miguel_es_pe` (Peruvian Spanish),`bel_pt_br` (Brazilian Portuguese),`david_es_es` (Castillian Spanish),`anna_ca` (Catalan). Please check the [Speech Center documentation](https://speechcenter.verbio.com/documentation) for an updated list of voices.


#### Sample rate


```
-s, --sample-rate arg
```

Sampling rate for the audio generation: 16000 (default: 16000).

#### Token


```
-t, --token arg 
```

Path to the authentication token file. This file needs to have a valid token in order for the Speech Center to work. 

In order for the client to work, the token argument is required in the following situations:

- The client will authenticate just by using the available token file. The file provided in this argument **needs to be a valid SpeechCenter JWT token so the transcription service can work**.
- The client will authenticate by providinf their client credentials through the `--client_id` and `--client_secret` program arguments. In this case **a token file must also be supplied even if it is a blank file**. Client will check file to see if it is a valid token, if it isn't it will refresh automatically the token and fill the file with a valid token. In this case, client_id and client_secret fields are also required.


#### Client id and secret


```
--client-id arg      Client id for token refresh (default: "")
```    

``` 
--client-secret arg  Client secret for token refresh (default: "")
``` 

`client-id` and `client-secret` fields are required for automatic token refreshal. The arguments need to be written inline with no quotes for each field. Please see the '**Automatically Refresh Service Token**' section above for specific details.

#### Host


```
-H, --host arg
```

URL of the host where the request will be sent. Main endpoints will be expanded as the product is deployed in different regions. Please use **us.speechcenter.verbio.com** as the host for CSR requests.






## Customer Credentials

All speech requests sent to Speech Center must have a valid authorization token in order for the system to successfully process the request. These tokens have a validity period of 1 hour and can be generated on demand using your customer credentials.

Your customer credentials can be retrieved through the [Speech Center Dashboard](https://dashboard.speechcenter.verbio.com) by logging in with the customer account supplied to you by Verbio.

### Authentication flow


To acquire a valid token submit an HTTP POST request to the authentication service at `https://auth.speechcenter.verbio.com:444`.

**Token expiration management**

![Authentication flow](img/OAuth_diagram.png)

As part of the JWT specification, we fully support the token expiration claim, so generated tokens will be valid for only a finite period of time of 1 hour to up to 1 day. It is the responsibility of the calling client to manage this token expiration time and, in a best case scenario, anticipate the refresh by a couple of minutes so the streaming session attempt never fails because of token expiration.

In order to refresh the token, the token refresh endpoint can be called with the same client_id and client_secret, and it will respond with a new JWT token with a renewed expiration time.


### Authentication API

**Method**
POST

**Resource**
```
/api/v1/token
```
**Request body**

```
{
 "client_id":"your_client_id",
 "client_secret":"your_client_secret"
}
```

**Response body**

Content type: application/json
Body:
```
{
 "access_token": "new_access_token",
 "expiration_time": 1678453328
}
```
*expiration_time field contains the expiration time for the JWT token so there is no need to decode the token on the client request to know the token expiration time.

Status codes:
```
HTTP 200: OK
HTTP 400: KO - Provided request format was invalid.
HTTP 401: KO - There was a problem with the provided credentials.
HTTP 5XX: KO - There was a problem with the server.
```

### Testing calls using curl

**Example request**
```
curl --header "Content-Type: application/json"   --request POST   --data '{"client_id":"YOUR_CLIENT_ID","client_secret":"YOUR_CLIENT_SECRET"}'   'https://auth.speechcenter.verbio.com:444/api/v1/token'
```

**Example response**

{
  "access_token": "EXAMPLE_ACCESS_TOKEN",
  "expiration_time": 1678453328
}

### Automatically Refresh Service Token

This repository optionally implements an automatic token update. To use it, you must specify your credentials (find them in the Client Credentials section of the [user dashboard](https://dashboard.speechcenter.verbio.com)).

You must also specify a token file, where the token will be stored and updated in case it is invalid or expired.

**Example**
```shell
python3 recognizer_stream.py --client-id="your-client-id" --client-secret="your-client-secret"
 --audio-file file.wav --topic GENERIC --language en-US --host us.speechcenter.verbio.com --token token.file --asr-version V1 --label project1

```

**WARNING**

Please note that due to a limitation of the command line argument parser, the `client_id` and `client_secret` arguments MUST be written in the following format since they might contain hyphens.

```text
--client-id="your-client-id"
           ^
```
> Note the usage of the `=` sign

Not defining the arguments like this will yield an error.

## Dashboard 

Speech Center features a dashboard at https://dashboard.speechcenter.verbio.com. 
Dashboard serves two main purposes:

1. Allow users to follow up historical usage data of the service, separated by projects or as a whole.
2. Allow users to retrieve client_id and client_secret credentials necessary to integrate with Speech Center, please look at authentication section to learn more about how this is performed. 


CHANGELOG.md
## To be released

### Features :tada:

### Bug fixes :bug:

### Removed :no_entry:

## 1.1.2 (2023-11-23)

### Patch (1 change)

- [Update version](csr/asr4-streaming@ebf1c88132d7f19b49507460f24dce9f7e87f651) ([merge request](csr/asr4-streaming!362))

## 1.0.6 (2023-11-13)

### Patch (5 changes)

- [Update w2v-engine versions to avoid conflicts with asr4-engine core](csr/asr4-streaming@e3b52a9827c8a1312d42a831bca926c416955666) ([merge request](csr/asr4-streaming!356))
- [[SC2-1152] Update W2V engine version to fix bugs](csr/asr4-streaming@54210b09a51801cf6f619a9a77a901a55353d1d5) ([merge request](csr/asr4-streaming!354))
- [Fix/Convert en-XX and es-XX language codes into w2v accepted language codes](csr/asr4-streaming@54d7c4344fb7afbf9fc26193d1530ef441aabfe9) ([merge request](csr/asr4-streaming!352))
- [[EEE-1141] Refactor/ Adjust timestamps and partial durations to total audio duration](csr/asr4-streaming@aa77d4a9db91ad4aeffed4e66a9c0e1d6aae3a3c) ([merge request](csr/asr4-streaming!348))
- [upgraded base image to 0.1.1](csr/asr4-streaming@e919739bee0ebc739cc48b11754d6a559b4158a0) ([merge request](csr/asr4-streaming!349))

## 1.0.5 (2023-11-09)

### Patch (4 changes)

- [[SC2-1152] Update W2V engine version to fix bugs](csr/asr4-streaming@54210b09a51801cf6f619a9a77a901a55353d1d5) ([merge request](csr/asr4-streaming!354))
- [Fix/Convert en-XX and es-XX language codes into w2v accepted language codes](csr/asr4-streaming@54d7c4344fb7afbf9fc26193d1530ef441aabfe9) ([merge request](csr/asr4-streaming!352))
- [[EEE-1141] Refactor/ Adjust timestamps and partial durations to total audio duration](csr/asr4-streaming@aa77d4a9db91ad4aeffed4e66a9c0e1d6aae3a3c) ([merge request](csr/asr4-streaming!348))
- [upgraded base image to 0.1.1](csr/asr4-streaming@e919739bee0ebc739cc48b11754d6a559b4158a0) ([merge request](csr/asr4-streaming!349))

## 1.0.4 (2023-10-11)

### Minor (1 change)

- [Update to version 1.4.0 of models](csr/asr4-streaming@7785d15df63de0e402c98c8b4eec4238262edced) ([merge request](csr/asr4-streaming!340))

### Patch (7 changes)

- [[Release 1.0.4] Fix argument processor mechanism in server.py and add test](csr/asr4-streaming@1060f0305ed2ef86954336ec4a2163d31e865187) ([merge request](csr/asr4-streaming!347))
- [Update asr4-engine to 0.10.2](csr/asr4-streaming@f30bc942e6cf7eedd0fb6ac471b7105a7cebe8c5) ([merge request](csr/asr4-streaming!346))
- [Changing all models to 1.0.3](csr/asr4-streaming@75e7dc795b9e596698fe4685e8df40a6cd74b740) ([merge request](csr/asr4-streaming!345))
- [[EEE-1051] Remove configuration that should be in the engine](csr/asr4-streaming@811a8f4338b28331a37ae6806cd8a98992787987) ([merge request](csr/asr4-streaming!343))
- [[SC2-1052] Refactor/ Close gRPC Channel on Internal Exception](csr/asr4-streaming@89d01899939776f7bddd72ffc7b33e81f388bb2f) ([merge request](csr/asr4-streaming!344))
- [Fix/ TRN File Generation](csr/asr4-streaming@9081c0286f600851fae00b9003bef7d531dc97d5) ([merge request](csr/asr4-streaming!342))
- [[EEE-1051] Remove unused arguments in server.py](csr/asr4-streaming@62c0715bd056c49313f5259133e451da45b99569) ([merge request](csr/asr4-streaming!341))

## 1.0.3 (2023-10-04)

### Patch (3 changes)

- [release/1.0.3](csr/asr4-streaming@66232d393766bedd6fceed61588d6024cf45693f) ([merge request](csr/asr4-streaming!337))
- [Fix/ Send Error Status Message before aborting a gRPC Channel](csr/asr4-streaming@fee6d15a950b8f44d716c5d9540466be8d94a421) ([merge request](csr/asr4-streaming!339))
- [[EEE-1050] Remove dependencies from W2V engine in client.py](csr/asr4-streaming@34904d10c9b04d38b796b6bba3cdddbebcbba4bf) ([merge request](csr/asr4-streaming!336))

## 1.0.0 (2023-09-15)

### Major (1 change)

- [[EEE-868] Feature/ Use asr4-engine Online Decoding API](csr/asr4@cebae65a774902bbcff830a37287294b7c571c4c) ([merge request](csr/asr4!302))

### Patch (7 changes)

- [release/1.0.0](csr/asr4@51f047a7d38605f4749496418f43c0d4a811ce60) ([merge request](csr/asr4!328))
- [[EEE-1085] Make client.py a bidirectional sender/reader](csr/asr4@7bc1f0adc1e064f782ffffbb64ea0292dc5e164c) ([merge request](csr/asr4!327))
- [Fix/ Streaming Functional Testing Latency Limit](csr/asr4@69eadc1761a292971d95ed14ceed389651a55079) ([merge request](csr/asr4!329))
- [[EEE-1049] Add loguru](csr/asr4@f086a3eefd6ff7aa40a0aea703ad1a8081904700) ([merge request](csr/asr4!324))
- [[EEE-945] Refactor/ gRPC Streaming API](csr/asr4@d758f34d21109f78cec24dcbe1101a1450de176f) ([merge request](csr/asr4!323))
- [Fix/ Code Coverage](csr/asr4@af243bd1a391dd780189c3bb19b34c3f65ea0ee2) ([merge request](csr/asr4!325))
- [[EEE-945] Refactor/ Remove gRPC Batch API](csr/asr4@ebb183ebea4c3fe5adfd6b1cfbe8ff9b6b10c165) ([merge request](csr/asr4!322))

## 0.7.8 (2023-09-04)

### Patch (9 changes)

- [Fix: Update release](csr/asr4@79aefbd9b8e9d6b0f9cfc4a74c24373c183adcd5) ([merge request](csr/asr4!321))
- [Increase grpc timeout from 3 to 4.](csr/asr4@9a6087350b2a5863f4c66e947faf54e1ac49b12d) ([merge request](csr/asr4!320))
- [Increase grpc timeout from 2 to 3](csr/asr4@8113df7cf98981c13e2f74c35476120cc59d0afc) ([merge request](csr/asr4!319))
- [Fix release 0.7.8](csr/asr4@0992924aebb9fe4179cb651b3dea810fcda74468) ([merge request](csr/asr4!318))
- [Fix some testing timeouts](csr/asr4@2234bda943ed50b66a702cc76d683d9da5f5dec5) ([merge request](csr/asr4!317))
- [Release 0.7.8](csr/asr4@b6a1c7afa54db9151e02744b5e7106f632f1aaa3) ([merge request](csr/asr4!316))
- [Fix/launch server ing gpu for integration tests](csr/asr4@a5864e1e12d717bfdc42b9d73044af40c45fd02c) ([merge request](csr/asr4!315))
- [Remove unused config file](csr/asr4@0f9419d8b00637087549537c9f24bf4b7b257a0c) ([merge request](csr/asr4!314))
- [[EEE-1044] fix streaming behaviour](csr/asr4@f1a54a32e3d7fd0fd5e1a6704ced4004dd3dcf91) ([merge request](csr/asr4!310))

## 0.6.7 (2023-07-14)

### Patch (4 changes)

- [Revert "now works compiled with torch 2.0.1"](csr/asr4@56495b1799b6bd5c474599920648b16c77fc75b5) ([merge request](csr/asr4!288))

cli-client/helpers/common.py
import math
import argparse
import logging
from helpers.speechcenterauth import SpeechCenterCredentials


class SynthesizerOptions:
    def __init__(self):
        self.token_file = None
        self.host = ""
        self.audio_file = None
        self.secure_channel = True
        self.audio_format = ''
        self.sample_rate: int = 0
        self.voice: str = None
        self.text: str = None
        self.client_id = None
        self.client_secret = None


def parse_credential_args(args, options):
    if args.client_id and not args.client_secret:
        raise argparse.ArgumentError(None, "If --client-id is specified, then --client-secret must also be specified.")
    elif args.client_secret and not args.client_id:
        raise argparse.ArgumentError(None, "If --client-secret is specified, then --client-id must also be specified.")
    options.client_id = args.client_id or None
    options.client_secret = args.client_secret or None


def check_commandline_values(args):
    if not args.text:
        logging.error("Synthesis text field needs to be non empty")
        raise ValueError("Synthesis text field needs to be non empty")


def parse_tts_command_line() -> SynthesizerOptions:
    options = SynthesizerOptions()
    parser = argparse.ArgumentParser(description='Perform speech synthesis on a given text')
    parser.add_argument('--text', '-T', help='Text to synthesize to audio', required=True)
    parser.add_argument(
        '--voice',
        '-v',
        choices=[
            'tommy_en_us',
            'miguel_es_pe',
            'bel_pt_br',
            'david_es_es',
            'anna_ca'],
        help='Voice to use for the synthesis',
        required=True)
    parser.add_argument('--sample-rate', '-s', type=int, choices=[16000], help='Output audio sample rate in Hz', default=16000)
    parser.add_argument('--format', '-f', choices=['wav', 'raw'], help='Output audio format', default='wav')
    parser.add_argument('--audio-file', '-a', help='Path to store the resulting audio', required=True)
    parser.add_argument('--token', '-t', help='File with the authentication token', required=True)
    parser.add_argument('--host', '-H', help='The URL of the host trying to reach', required=True)
    parser.add_argument('--not-secure', '-S', help='Do not use a secure channel. Used for internal testing.',
                        required=False, default=True, dest='secure', action='store_false')

    credentialGroup = parser.add_argument_group(
        'credentials',
        '''[OPTIONAL] Client authentication credentials used to refresh the token.
        You can find your credentials on the dashboard at https://dashboard.speechcenter.verbio.com/access-token''')
    credentialGroup.add_argument('--client-id', help='Client id for authentication. MUST be written as --client-id=CLIENT_ID')
    credentialGroup.add_argument('--client-secret', help='Client secret for authentication. MUST be written as --client-secret=CLIENT_SECRET')

    args = parser.parse_args()
    check_commandline_values(args)
    parse_credential_args(args, options)

    options.token_file = args.token
    options.host = args.host
    options.audio_file = args.audio_file
    options.secure_channel = args.secure
    options.audio_format = args.format
    options.text = args.text
    options.voice = args.voice
    options.sample_rate = args.sample_rate

    return options


def retrieve_token(options: SynthesizerOptions) -> str:
    if options.client_id:
        return SpeechCenterCredentials.get_refreshed_token(options.client_id, options.client_secret, options.token_file)
    else:
        return SpeechCenterCredentials.read_token(token_file=options.token_file)


class RecognizerOptions:
    def __init__(self):
        self.token_file = None
        self.host = ""
        self.audio_file = None
        self.topic = None
        self.language = 'en-US'
        self.secure_channel = True
        self.diarization = False
        self.formatting = False
        self.inactivity_timeout = False
        self.asr_version = None
        self.label = None
        self.client_id = None
        self.client_secret = None

    def check(self):
        if self.topic is None:
            raise Exception("You must provide a least a topic")



def parse_csr_commandline() -> RecognizerOptions:
    options = RecognizerOptions()
    parser = argparse.ArgumentParser(description='Perform speech recognition on an audio file')
    parser.add_argument('--audio-file', '-a', help='Path to a .wav audio in 8kHz and PCM16 encoding', required=True)
    topicGroup = parser.add_mutually_exclusive_group(required=True)
    topicGroup.add_argument('--topic', '-T', choices=['GENERIC', 'TELCO', 'BANKING', 'INSURANCE'], help='A valid topic')
    parser.add_argument(
        '--language',
        '-l',
        choices=[
            'en',
            'en-US',
            'en-GB',
            'pt-BR',
            'es',
            'es-419',
            'tr',
            'ja',
            'fr',
            'fr-CA',
            'de',
            'it'],
        help='A Language ID (default: ' + options.language + ')',
        default=options.language)
    parser.add_argument('--token', '-t', help='File with the authentication token', required=True)
    parser.add_argument('--host', '-H', help='The URL of the host trying to reach (default: ' + options.host + ')', required=True)
    parser.add_argument('--not-secure', '-S', help='Do not use a secure channel. Used for internal testing.',
                        required=False, default=True, dest='secure', action='store_false')
    parser.add_argument('--diarization', '-d', help='', required=False, default=False, action='store_false')
    parser.add_argument('--formatting', '-f', help='', required=False, default=False, action='store_false')
    parser.add_argument('--inactivity-timeout', '-i', help='Time for stream inactivity after the first valid response', required=False, default=5.0)
    parser.add_argument('--asr-version', choices=['V1', 'V2'], help='Selectable asr version', required=True)
    parser.add_argument('--label', help='"Label for the request', required=False, default="")

    credentialGroup = parser.add_argument_group(
        'credentials',
        '''[OPTIONAL] Client authentication credentials used to refresh the token.
        You can find your credentials on the dashboard at https://dashboard.speechcenter.verbio.com/access-token''')
    credentialGroup.add_argument('--client-id', help='Client id for authentication. MUST be written as --client-id=CLIENT_ID')
    credentialGroup.add_argument('--client-secret', help='Client secret for authentication. MUST be written as --client-secret=CLIENT_SECRET')

    args = parser.parse_args()
    parse_credential_args(args, options)

    options.token_file = args.token
    options.host = args.host
    options.audio_file = args.audio_file
    options.topic = args.topic
    options.language = args.language
    options.secure_channel = args.secure
    options.formatting = args.formatting
    options.diarization = args.diarization
    options.inactivity_timeout = float(args.inactivity_timeout)
    options.asr_version = args.asr_version
    options.label = args.label

    return options


def split_audio(audio: bytes, chunk_size: int = 20000):
    audio_length = len(audio)
    chunk_count = math.ceil(audio_length / chunk_size)
    logging.info("Dividing audio of length " + str(audio_length) + " into " + str(chunk_count) + " of size " + str(chunk_size) + "...")
    if chunk_count > 1:
        for i in range(chunk_count):
            start = i * chunk_size
            end = min((i + 1) * chunk_size, audio_length)
            logging.info("Audio chunk #" + str(i) + " sliced as " + str(start) + ":" + str(end))
            yield audio[start:end]
    else:
        yield audio
